### PR TITLE
New endpoints with cleaner extraction code plus tests

### PR DESCRIPTION
**Pull request by MartinBentley**
There is a new course controller with a more granular fetch for course/s.
Also the Organisations controller has been refactored.
All data extraction are now using raw sql in the context object

### CODE REVIEW

File: src/ManageCourses.Domain/DatabaseAccess/ManageCoursesDbContext.cs [hunk 1]
```diff
@@ -243,12 +244,63 @@ public void AddMcUser(McUser user)
         {
             McUsers.Add(user);
         }
+        public List<UcasCourse> GetUcasCourseRecordsByUcasCode(string instCode, string ucasCode, string email)
+        {
+            var ucasCourses = UcasCourses.FromSql(
+                    $"select c.* from ucas_course c " +
```
Comment (timabell) [inline src/ManageCourses.Domain/DatabaseAccess/ManageCoursesDbContext.cs]: Don't use `*`. It's fragile and hard to troubleshoot.
I don't think using raw sql is advisable unless there's an exceptionally good reason for it. We lose all of the strong type checking and ability to find usages etc that we gain from sticking to c# & linq.

**Comment by timabell** [comment]
I think using raw sql should be avoided if at all possible. See comments in code.

File: tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs [hunk 2]
```diff
@@ -250,5 +253,264 @@ public void GetCoursesForUser_isNull(string email, string orgId)
...
+            const int numCourses = 6;
+            LoadData(TestUserEmail_1, numOrgs, numCourses);
+            var orgList = Subject.GetOrganisationsForUser(TestUserEmail_1).ToList();
+            Assert.IsTrue(orgList.Count == numOrgs);
+            Assert.IsTrue(orgList.All(c => c.TotalCourses == numCourses));
+
+            foreach (var org in orgList)//we have a valist list of data
+            {
+                var result = Subject.GetCourses(TestUserEmail_1, org.UcasCode);//get the course for each org
+                Assert.IsTrue(result != null && result.Courses.Count == numCourses);
```
Comment (defong) [inline tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs]: Can we start breaking it out as 
`Assert.IsNotNull(result)` <= that can be obmitted as the below line actual needs it
`Assert.AreEqual(numCourses , result.Courses.Count)`

Comment (MartinBentley) [reply tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs]: I agree. Much cleaner

File: tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs [hunk 3]
```diff
@@ -250,5 +253,264 @@ public void GetCoursesForUser_isNull(string email, string orgId)
...
+        [Test]
+        [TestCase(null, "")]
+        [TestCase("", null)]
+        [TestCase(null, null)]
+        [TestCase("", "")]
+        [TestCase("     ", "    ")]
+        [TestCase("anyone@anywher.com", "")]
+        [TestCase("", "ucasCode")]
+        [TestCase("anyon@anywhere.com", "ABC")]
+        public void GetCoursesWithInvalidUserAndInvalidUcasCodeShouldNotReturnNoData(string email, string ucasCode)
```
[the remark this answers is not in the document]
Comment (MartinBentley) [reply tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs]: Checked........... they are being run

File: src/ManageCourses.ApiClient/Properties/PublishProfiles/FolderProfile.pubxml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+<!--
+https://go.microsoft.com/fwlink/?LinkID=208121. 
+-->
+<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <PublishProtocol>FileSystem</PublishProtocol>
+    <Configuration>Release</Configuration>
+    <Platform>Any CPU</Platform>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <PublishDir>C:\Martin\Repos\Publish</PublishDir>
```
[the remark this answers is not in the document]
Comment (MartinBentley) [reply src/ManageCourses.ApiClient/Properties/PublishProfiles/FolderProfile.pubxml]: oops. Removed!

File: src/ManageCourses.Api/Controllers/CoursesController.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            _dataService = dataService;
+        }
+
+        /// <summary>
+        /// Gets a course by institution Ucas code
+        /// </summary>
+        /// <returns>a single course</returns>
+        [Authorize]
+        [HttpGet]
+        [Route("{instCode}/{ucasCode}")]
```
[the remark this answers is not in the document]
Comment (MartinBentley) [reply src/ManageCourses.Api/Controllers/CoursesController.cs]: Done

File: tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (defong) [reply tests/ManageCourses.Tests/DbIntegration/DataServiceTests.cs]: Nice one

File: src/ManageCourses.Domain/DatabaseAccess/ManageCoursesDbContext.cs (hunk 1 above)
Comment (timabell) [reply src/ManageCourses.Domain/DatabaseAccess/ManageCoursesDbContext.cs]: :+1:

**Pull request merged**